poplar: add initial poplar.mk

igoropaniuk opened the pull request:
Add initial poplar.mk, not integrated with buildroot yet.

We are using Poplar board intensively in GL, so why not to introduce the same setup the same for Hikey to avoid painful manual bring-up. 

Includes `flash-help` target with all steps about flashing partitions to the destination device.
Relates to: https://github.com/OP-TEE/manifest/pull/115/files

`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

vchong commented:
Would be nice to add `CFG_DRAM_SIZE_GB ?= 2` to `OPTEE_OS_COMMON_FLAGS`.

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+ROOTFS_BIN		?= linaro-stretch-developer-20170914-80.tar.gz
+ROOTFS_URL		?= http://download.djjproject.com/poplar/$(ROOTFS_BIN)
+
+PKG_OPTEE_VERSION	?= 1.0.0
```
Why not `?= $(shell cd $(OPTEE_OS_PATH) && git describe)-0`?

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+################################################################################
+# Folders
+################################################################################
+$(OUT_PATH):
```
Who calls this target?

igoropaniuk replied:
will delete it

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+.PHONY: optee-test-clean
+optee-test-clean:
+	$(MAKE) -C $(OPTEE_TEST_PATH) clean
+
```
Can't you just depend on targets in `common.mk` for `optee-client`, `optee-client-clean`, `optee-test` and `optee-test-clean`?

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+.PHONY: arm-tf
+arm-tf: u-boot optee-os
+	$(MAKE) -C $(ARM_TF_PATH) \
+		CROSS_COMPILE="$(AARCH64_CROSS_COMPILE)" \
```
nit: `CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)" \`

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+	$(MAKE) -C $(ARM_TF_PATH) \
+		CROSS_COMPILE="$(AARCH64_CROSS_COMPILE)" \
+		all fip \
+		DEBUG=1 \
```
`DEBUG=$(DEBUG)`

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+################################################################################
+.PHONY: deb-package
+deb-package: optee-test optee-client
+	mkdir -p $(OUT_PATH)/debpkg
```
Can remove

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+.PHONY: deb-package
+deb-package: optee-test optee-client
+	mkdir -p $(OUT_PATH)/debpkg
+	mkdir -p $(PKG_PATH)
```
Can remove

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+	@echo ""
+	@echo "   => setenv ipaddr 192.168.0.2"
+	@echo "   => setenv netmask 255.255.255.0"
+	@echo "   => setenv serverip <your_host_ip_address>"
```
Maybe helpful to mention that host ip is set to 192.168.0.3 as to example below.

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+	@echo ""
+	@echo "9. Upload deb package using SCP (run on your host PC):"
+	@echo ""
+	@echo "   $$ scp optee_1.0.0.deb linaro@192.168.0.2:/tmp"
```
`optee_${PKG_OPTEE_VERSION}.deb`

igoropaniuk commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+OPTEE_BIN_EXTRA2	?= $(OPTEE_PATH)/out/arm/core/tee-pageable_v2.bin
+
+ROOTFS_BIN		?= linaro-stretch-developer-20170914-80.tar.gz
+ROOTFS_URL		?= http://download.djjproject.com/poplar/$(ROOTFS_BIN)
```
The problem with latest rootfs that is hosted on snapshots.linaro.org, is that it is flashed with errors to mmc (seems that poplar tools create wrong partitions images). That's why I started using the same image, that was mentioned in the official manual for bringing up debian on poplar (but, unfortunately, it's doesn't exist on snapshots.linaro.org anymore).

I don't remember exact error message, but will

igoropaniuk replied:
So the problem occurs when `unzip` is called for  recovery_files/partition3.2-of-8.gz.

```
ETH1: PHY(phyaddr=3, rgmii) link UP: DUPLEX=FULL : SPEED=100M
Using gmac1 device
TFTP from server 10.103.106.70; our IP address is 10.103.106.64
Filename 'recovery_files/partition3.2-of-8.gz'.
Load address: 0x8000000
Loading: #################################################################

done
Bytes transferred = 138279169 (83df901 hex)
Error: inflate() returned -3

MMC write: dev # 0, block # 794624, count 524288 ... 524288 blocks written: OK

```

I guess the root cause could be an insufficient amount of RAM space to store un-gziped partition (its gziped size is 132M )

```
out/recovery_files$ ls -lah
total 371M
drwxrwxrwx 2 igoropaniuk igoropaniuk 4,0K чер  5 15:52 .
drwxrwxr-x 4 igoropaniuk igoropaniuk 4,0K чер  5 15:33 ..
-rwxrwxrwx 1 igoropaniuk igoropaniuk 2,0M чер  5 15:31 fastboot.bin
-rwxrwxrwx 1 igoropaniuk igoropaniuk  645 чер  5 15:33 install
-rwxrwxrwx 1 igoropaniuk igoropaniuk  169 чер  5 15:31 install-layout
-rwxrwxrwx 1 igoropaniuk igoropaniuk  241 чер  5 15:31 install-layout.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk  168 чер  5 15:32 install-partition1
-rwxrwxrwx 1 igoropaniuk igoropaniuk  240 чер  5 15:32 install-partition1.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk  167 чер  5 15:32 install-partition2
-rwxrwxrwx 1 igoropaniuk igoropaniuk  239 чер  5 15:32 install-partition2.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk 1,2K чер  5 15:33 install-partition3
-rwxrwxrwx 1 igoropaniuk igoropaniuk 1,3K чер  5 15:33 install-partition3.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk  717 чер  5 15:33 install.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk   86 чер  5 15:31 mbr.bin.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 396K чер  5 15:32 partition1.1-of-1.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 6,1M чер  5 15:32 partition2.1-of-1.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  42M чер  5 15:32 partition3.1-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 132M чер  5 15:32 partition3.2-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  93M чер  5 15:32 partition3.3-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  77M чер  5 15:33 partition3.4-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  20M чер  5 15:33 partition3.5-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 255K чер  5 15:33 partition3.6-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 255K чер  5 15:33 partition3.7-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 161K чер  5 15:33 partition3.8-of-8.gz

```

igoropaniuk replied:
```
$ git diff
diff --git a/poplar_recovery_builder.sh b/poplar_recovery_builder.sh
index e4493b4..c9dd0e2 100755
--- a/poplar_recovery_builder.sh
+++ b/poplar_recovery_builder.sh
@@ -18,7 +18,7 @@ EMMC_SIZE=14942208    # 7296 MB in sectors (not hex)
 CHUNK_SIZE=524288      # Partition image chuck size in sectors (not hex)
 
 IN_ADDR=0x08000000     # Buffer address for compressed data in from USB (hex)
-OUT_ADDR=0x10000000    # Buffer address for uncompressed data for MMC (hex)
+OUT_ADDR=0x11000000    # Buffer address for uncompressed data for MMC (hex)
 SUB_ADDR=0x07800000    # Buffer address for sub-installer scripts
 
 EMMC_DEV=/dev/mmcblk0  # Linux path to main eMMC device on target
@@ -118,7 +118,7 @@ function parseargs() {
        PARTS=$1
        shift
 
-       LOAD_COMMAND=tftp
+       LOAD_COMMAND=tftpboot
        INPUT_FILES=""
 
        while [ $# -gt 0 ]; do

```

igoropaniuk replied:
https://github.com/96boards-poplar/poplar-tools/pull/4

igoropaniuk commented on poplar.mk:
```diff
@@ -0,0 +1,304 @@
+OPTEE_BIN_EXTRA1	?= $(OPTEE_PATH)/out/arm/core/tee-pager_v2.bin
+OPTEE_BIN_EXTRA2	?= $(OPTEE_PATH)/out/arm/core/tee-pageable_v2.bin
+
+ROOTFS_BIN		?= linaro-stretch-developer-20170914-80.tar.gz
```
@vchong we have to specify full filename, because it's fetched with wget, check `rootfs` make target.

igoropaniuk commented:
@vchong @jbech-linaro the problem of integration with buildroot is that we depend on `poplar_recovery_builder.sh` script, which consumes rootfs in tar.gz format and then creates proper partition images and generates U-boot shell scripts, that simplify flashing of these partitions from U-boot shell.

As I understand, it will require some effort to create something similar on top of buildroot.

My suggestion is to merge what we have now (after addressing issues), and then proceed with buildroot integration in a separate PR.

jbech-linaro commented:
> @vchong @jbech-linaro the problem of integration with buildroot is that we depend on poplar_recovery_builder.sh
> ...
> My suggestion is to merge what we have now (after addressing issues), and then proceed with buildroot integration in a separate PR.

That's fine, there no strict rule that we shall run Buildroot on everything either. It's certainly good if we can do that and something we strive for, since then we have a common way of doing things. But right now, this is new stuff, so I'm fine with your suggestions here.

igoropaniuk commented:
All comments addressed

vchong commented on poplar.mk:
```diff
@@ -0,0 +1,286 @@
+ROOTFS_BIN		?= linaro-stretch-developer-*.tar.gz
+ROOTFS_URL		?= http://snapshots.linaro.org/debian/images/stretch/developer-arm64/latest/$(ROOTFS_BIN)
+
+PKG_OPTEE_VERSION	?= $(shell cd $(OPTEE_OS_PATH) && git describe)
```
Need to end with `-0`. See https://github.com/OP-TEE/build/pull/154#discussion_r121323471 for details.

Other than that, `Reviewed-by: Victor Chong <victor.chong@linaro.org>`

igoropaniuk replied:
fixed, thanks!

igoropaniuk commented:
tag applied, rebased and ready for merge

igoropaniuk commented:
BTW, https://github.com/96boards-poplar/poplar-tools/pull/4 was also merged
